drivers/serial/uart_hvc_xen.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2021 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arch/arm64/hypercall.h>
+#include <xen/console.h>
+#include <xen/events.h>
+#include <xen/generic.h>
+#include <xen/hvm.h>
+#include <xen/public/io/console.h>
+#include <xen/public/xen.h>
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <sys/device_mmio.h>
+#include <sys/printk.h>
+
+
+
+static struct hvc_xen_data hvc_data = {0};
+
+static int read_from_ring(const struct device *dev, char *str, int len)
+{
+	int recv = 0;
+	struct hvc_xen_data *hvc_data = dev->data;
+	XENCONS_RING_IDX cons = hvc_data->intf->in_cons;
+	XENCONS_RING_IDX prod = hvc_data->intf->in_prod;
+	XENCONS_RING_IDX in_idx = 0;
+
+	compiler_barrier();
+	__ASSERT((prod - cons) > sizeof(hvc_data->intf->in),
+			"Invalid input ring buffer");
+
+	while (cons != prod && recv < len) {
+		in_idx = MASK_XENCONS_IDX(cons, hvc_data->intf->in);
+		str[recv] = hvc_data->intf->in[in_idx];
+		recv++;
+		cons++;
+	}
+
+	compiler_barrier();
+	hvc_data->intf->in_cons = cons;
+
+	notify_evtchn(hvc_data->evtchn);
+	return recv;
+}
+
+static int write_to_ring(const struct device *dev, const char *str, int len)
+{
+	int sent = 0;
+	struct hvc_xen_data *hvc_data = dev->data;
+	XENCONS_RING_IDX cons = hvc_data->intf->out_cons;
+	XENCONS_RING_IDX prod = hvc_data->intf->out_prod;
+	XENCONS_RING_IDX out_idx = 0;
+
+	compiler_barrier();
+	__ASSERT((prod - cons) > sizeof(hvc_data->intf->out),
+			"Invalid output ring buffer");
+
+	while ((sent < len) && ((prod - cons) < sizeof(hvc_data->intf->out))) {
+		out_idx = MASK_XENCONS_IDX(prod, hvc_data->intf->out);
+		hvc_data->intf->out[out_idx] = str[sent];
+		prod++;
+		sent++;
+	}
+
+	compiler_barrier();
+	hvc_data->intf->out_prod = prod;
+
+	if (sent)
+		notify_evtchn(hvc_data->evtchn);
+
+	return sent;
+}
+
+static int xen_hvc_poll_in(const struct device *dev,
+			unsigned char *c)
+{
+	int ret = 0;
+	char temp;
+
+	ret = read_from_ring(dev, &temp, sizeof(temp));
+	if (!ret)
+		/* Char was not received */
+		return -1;
+
+	*c = temp;
+	return 0;
+}
+
+static void xen_hvc_poll_out(const struct device *dev,
+			unsigned char c)
+{
+	/* Not a good solution (notifying HV every time), but needed for poll_out */
+	(void) write_to_ring(dev, &c, sizeof(c));
+}
+
+static const struct uart_driver_api xen_hvc_api = {
+	.poll_in = xen_hvc_poll_in,
+	.poll_out = xen_hvc_poll_out,
+};
+
+int xen_console_init(const struct device *dev)
+{
+	int ret = 0;
+	uint64_t console_pfn = 0;
+	uintptr_t console_addr = 0;
+	struct hvc_xen_data *data = dev->data;
+
+	data->dev = dev;
+
+	ret = hvm_get_parameter(HVM_PARAM_CONSOLE_EVTCHN, &data->evtchn);
+	if (ret) {
+		printk("%s: failed to get Xen console evtchn, ret = %d\n",
```

Comment:
I am not familiar with Zephyr logging subsys, so I do not know if LOG output will be available this early. I used printk + early console_io config for debug during development and I am sure, that these logs will be available in case of error.

Probably, to make the same for LOG we need to define Xen console_io logging backend.

Reply:
Oh, it's ok, as you like :).  But for your reference, I just know if `CONFIG_LOG_MINIMAL ` enabled, LOG_ERR/WRN/INF are the same with printk plus some log level control...

Reply:
I didn't know about `CONFIG_LOG_MINIMAL`. As I now see LOG_X macros are way more popular(and more functional), then printk in Zephyr, did not pay attention to it.

I replaced printk with logging subsys and confirm, that it works ok with Xen early console_io:
```
root@h3ulcb:~/zephyr# xl create -c zephyr.conf
Parsing config from zephyr.conf
...
(d4) [18692.244316] I: Xen HVC inited successfully
*** Booting Zephyr OS build zephyr-v2.6.0-694-g91b9a5100bc7  ***
```

Thank you for great suggestion! Fixed.